### ISSUE TITLE
Create a context manager to enable InferenceMode in python frontend

### ISSUE THREAD

**Issue by ailzhang** [pytorch/pytorch#56608]
Since we're mostly done in InferenceMode on C++ end, it's possible that we add a context manager to enable InferenceMode in python frontend. This will give perf boost for python inference workload compare to no_grad guard we currently have (also with a bit more limitations). 
1. Context manager might look similar to https://github.com/pytorch/pytorch/blob/63997db6ec00d0cf60b30b1a6ff5e9d1bcf4b22a/torch/autograd/grad_mode.py#L84 and here're a few tests to cover in python end. https://github.com/pytorch/pytorch/blob/63997db6ec00d0cf60b30b1a6ff5e9d1bcf4b22a/test/cpp/api/inference_mode.cpp
2. We also need to add documentation explaining to users when to use no_grad guard and inference_mode guard. This can wait a bit since I'm planning to write a doc for InferenceMode API and we can likely link that in this section as well. 
cc: @albanD 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

**Comment by ailzhang** [comment]
@cpuhrsch is working on nested tensor which only works for inference workload. A python wrapper is very helpful to unblock the workstream.

**Comment by cpuhrsch** [comment]
This is also useful a situations in torchaudio where supporting a derivative (or higher derivatives) can have [influence on the forward performance](https://github.com/pytorch/audio/pull/1441).

**Comment by albanD** [comment]
I think this is different? In that case, you can check if `GradMode::is_enabled()` to know if you need to use a differentiable implementation or not.

**Comment by ailzhang** [comment]
Yea the torchaudio case might be a bit different. InferenceMode should only be used in the case that you know backward will never be called.

**Comment by cpuhrsch** [comment]
Yes, I agree, it's not exactly the same thing. But it showcases a scenario where higher order derivatives increase the runtime for all backward passes, even if the higher order wasn't needed. A guard or setting to restrict autograd to a certain order could allow us to guard that extra compute.

**Comment by ezyang** [comment]
It is certainly true that not knowing how many levels of derivative you are going to do can negatively impact performance. I tried to find an issue for this, but I couldn't find one. @cpuhrsch maybe it would be good to file a fresh issue for this.

**Comment by albanD** [comment]
You can already use `torch.is_grad_enabled()` (or the corresponding c++ API) to see if you are in a context where you might end up creating order + 1 derivative.

**Comment by ezyang** [comment]
@albanD but there is no way to say, "I guarantee that I am only going to do two differentiations" at the point when you are doing the initial forwards

**Comment by albanD** [comment]
Indeed, but I'm not sure if we actually have use cases that need that.
A separate issue to discuss this sounds good in any case.

**Comment by cpuhrsch** [comment]
I'll add to my todo list to write up a separate issue. One use case is [higher order derivatives for filters](https://github.com/pytorch/audio/pull/1441) (see the timings).

**Comment by albanD** [comment]
Note that the doc part of this should also explain what is `nn.Module.eval()` and what is the link with these two guards.
